api/scripts/arborescence-monitoring/time-series.js

```diff
@@ -0,0 +1,64 @@
+import { assertNotNullOrUndefined } from '../../src/shared/domain/models/asserts.js';
+
+const MAXIMUM_NUMBER_OF_POINTS_THRESHOLD = 28;
+
+const PURGE_POINT_INDEX = 10;
+
+const NB_POINTS_TO_PURGE = 1;
+
+const NULL_OR_UNDEFINED_POINTS_ERROR_MESSAGE = 'Time series should have a non null/undefined array of points';
+
+export class TimeSeries {
+  #points;
+  constructor(points) {
+    assertNotNullOrUndefined(points, NULL_OR_UNDEFINED_POINTS_ERROR_MESSAGE);
+    points.sort((pointA, pointB) => new Date(pointA.x).getTime() - new Date(pointB.x).getTime());
+    this.#points = points;
+  }
+
+  size() {
+    return this.#points.length;
+  }
+
+  get(pointIndex) {
+    return this.#points[pointIndex];
+  }
+
+  toString() {
+    return JSON.stringify(this.#points);
+  }
+
+  add({ x, y }) {
+    if (this.#sameValueThanTheLast(y)) {
+      return this.#replaceLastPointByNewPoint({ x, y });
+    } else {
+      this.#points.push({ x, y });
+      if (this.#exceedMaxNumberOfPoints()) {
+        this.#purgeOnePoint();
+      }
+      return new TimeSeries(this.#points);
```

Comment:
Ce renvoi d'une nouvelle instance de la classe m'interroge, sur une fonction qui est un verbe de mutation de l'instance actuelle.

Quelle est la raison de la création d'une nouvelle instance ? Si c'est pour le sort du constructeur je proposerai plutôt de sortir le sort et de ré-appliquer à `this.#points` le sort en fin de cette fonction

Reply:
C'est une technique/style de code basée sur de l'immutabilité (du fonctionnelle), où les fonctions publiques qui mutent l'état, retournent une nouvelle instance dont les invariants sont garanties par le constructeur, c'est très pratique;).

Reply:
Je comprends, et je connais la technique. Par contre c'est pas une pratique dans le code de Pix, et ça rompt un peu les habitudes de la codebase actuelle (invariant + fonction immutables).

J'adore la FP, mais à voir si on veut commencer à introduire cela full balle alors que les devs sont pas habitués.

C'est d'ailleurs ce qui m'a fait réagir en premier lieu, tellement pas l'habitude que Pix soit FP, que j'ai commenté.

En résumé : ça me dérange pas mais c'est disruptif par rapport à notre codebase actuelle.

Reply:
Je comprends mieux, mais pour un script destiné à disparaitre je pense qu'il y a peu de risque que ça pose un problème de maintenance.